In [2]:
import pandas as pd
# conda install pyjanitor -c conda-forge
from janitor import clean_names

# MVP

## Joins

### Question 1

Read in all 4 credit card transaction datasets and clean up the column names.

- CardBase.csv
- CustomerBase.csv
- FraudBase.csv
- TransactionBase.csv

In [3]:
card_base = pd.read_csv("data/CardBase.csv").clean_names()

In [9]:
card_base

card_number card_family  credit_limit  cust_id
0    8638-5407-3631-8196     Premium        530000  CC67088
1    7106-4239-7093-1515        Gold         18000  CC12076
2    6492-5655-8241-3530     Premium        596000  CC97173
3    2868-5606-5152-5706        Gold         27000  CC55858
4    1438-6906-2509-8219    Platinum        142000  CC90518
..                   ...         ...           ...      ...
495  4708-4407-9601-6022     Premium        691000  CC64993
496  6042-2856-7280-2925        Gold         33000  CC26787
497  8706-3809-5167-3899     Premium        144000  CC32532
498  9540-8558-5897-5046     Premium        830000  CC90246
499  5587-7265-8118-7718        Gold         24000  CC37803

[500 rows x 4 columns]

In [5]:
customer_base = pd.read_csv("data/CustomerBase.csv").clean_names()

In [10]:
customer_base

cust_id  age customer_segment customer_vintage_group
0     CC25034   35          Diamond                    VG1
1     CC59625   22          Diamond                    VG1
2     CC69314   33          Diamond                    VG1
3     CC67036   46          Diamond                    VG1
4     CC25597   44          Diamond                    VG1
...       ...  ...              ...                    ...
5669  CC53430   27         Platinum                    VG2
5670  CC31823   33         Platinum                    VG2
5671  CC15335   28         Platinum                    VG2
5672  CC81556   26         Platinum                    VG2
5673  CC38634   21         Platinum                    VG2

[5674 rows x 4 columns]

In [6]:
fraud_base = pd.read_csv("data/FraudBase.csv").clean_names()

In [12]:
fraud_base

transaction_id  fraud_flag
0     CTID50558449           1
1     CTID55936882           1
2     CTID63762180           1
3     CTID76723439           1
4     CTID21246201           1
..             ...         ...
104   CTID57993591           1
105   CTID91108283           1
106   CTID30494187           1
107   CTID51301522           1
108   CTID49517337           1

[109 rows x 2 columns]

In [7]:
transaction_base = pd.read_csv("data/TransactionBase.csv").clean_names()

In [13]:
transaction_base

transaction_id transaction_date       credit_card_id  transaction_value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     transaction_segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

### Question 2

Join the data containing card details and customer details by customer id, so that all records of card details and any matching records in customer details are kept.

In [11]:
card_base.merge(customer_base, how='left', on='cust_id')

card_number card_family  credit_limit  cust_id  age  \
0    8638-5407-3631-8196     Premium        530000  CC67088   27   
1    7106-4239-7093-1515        Gold         18000  CC12076   48   
2    6492-5655-8241-3530     Premium        596000  CC97173   22   
3    2868-5606-5152-5706        Gold         27000  CC55858   30   
4    1438-6906-2509-8219    Platinum        142000  CC90518   50   
..                   ...         ...           ...      ...  ...   
495  4708-4407-9601-6022     Premium        691000  CC64993   34   
496  6042-2856-7280-2925        Gold         33000  CC26787   46   
497  8706-3809-5167-3899     Premium        144000  CC32532   28   
498  9540-8558-5897-5046     Premium        830000  CC90246   35   
499  5587-7265-8118-7718        Gold         24000  CC37803   21   

    customer_segment customer_vintage_group  
0            Diamond                    VG1  
1               Gold                    VG3  
2            Diamond                    VG1  
3            Diamond                    VG1  
4           Platinum                    VG2  
..               ...                    ...  
495          Diamond                    VG1  
496         Platinum                    VG2  
497         Platinum                    VG2  
498             Gold                    VG3  
499          Diamond                    VG1  

[500 rows x 7 columns]

### Question 3

Join the data containing fraud details with transaction details so all rows of both tables are kept. What does the resulting row number tell you?

In [15]:
fraud_base.merge(transaction_base, how='outer', on='transaction_id')

transaction_id  fraud_flag transaction_date       credit_card_id  \
0      CTID50558449         1.0         6-May-16  4246-1369-3659-8804   
1      CTID55936882         1.0        29-Nov-16  1336-9200-1264-2551   
2      CTID63762180         1.0         5-Dec-16  8528-6154-7390-5081   
3      CTID76723439         1.0        15-Sep-16  7908-2695-7391-7499   
4      CTID21246201         1.0        29-Feb-16  2524-4184-5908-6750   
...             ...         ...              ...                  ...   
9995   CTID25037573         NaN        10-Jun-16  9157-2802-8374-1145   
9996   CTID43832298         NaN        23-May-16  7416-4529-6690-5703   
9997   CTID56532072         NaN         6-Aug-16  6699-2639-4522-6219   
9998   CTID88101446         NaN        10-Oct-16  8341-5263-4582-7396   
9999   CTID64730271         NaN        18-Jun-16  1947-8602-1695-7503   

      transaction_value transaction_segment  
0                  6984               SEG23  
1                 34367               SEG21  
2                 44550               SEG15  
3                 48275               SEG16  
4                 35751               SEG25  
...                 ...                 ...  
9995              44280               SEG18  
9996              27163               SEG12  
9997              36424               SEG13  
9998               8464               SEG14  
9999               7606               SEG15  

[10000 rows x 6 columns]

It seems that every row of `fraud_base` had a match in `transaction_base`, as both `transaction_base` and the joined dataset have 10000 rows.

### Question 4

**[Harder]** Join the data containing card details with transaction details so rows from the first which have matching ones in 
the second are returned.

Hint: `semi` is not an allowed option for `how` in `.merge()`. You'll need to think about the process of a semi 
join and then perform each step.

In [16]:
card_base.merge(transaction_base, how='inner', left_on='card_number', right_on='credit_card_id')

card_number card_family  credit_limit  cust_id transaction_id  \
0     8638-5407-3631-8196     Premium        530000  CC67088   CTID85973577   
1     8638-5407-3631-8196     Premium        530000  CC67088   CTID13215166   
2     8638-5407-3631-8196     Premium        530000  CC67088   CTID41016745   
3     8638-5407-3631-8196     Premium        530000  CC67088   CTID82512584   
4     8638-5407-3631-8196     Premium        530000  CC67088   CTID97222954   
...                   ...         ...           ...      ...            ...   
9995  5587-7265-8118-7718        Gold         24000  CC37803   CTID76550379   
9996  5587-7265-8118-7718        Gold         24000  CC37803   CTID58221219   
9997  5587-7265-8118-7718        Gold         24000  CC37803   CTID73252195   
9998  5587-7265-8118-7718        Gold         24000  CC37803   CTID72271256   
9999  5587-7265-8118-7718        Gold         24000  CC37803   CTID81265917   

     transaction_date       credit_card_id  transaction_value  \
0           13-Feb-16  8638-5407-3631-8196              31592   
1           30-Oct-16  8638-5407-3631-8196               5894   
2            9-May-16  8638-5407-3631-8196              20925   
3           31-Aug-16  8638-5407-3631-8196              36556   
4           25-Nov-16  8638-5407-3631-8196              36715   
...               ...                  ...                ...   
9995        11-Oct-16  5587-7265-8118-7718              23298   
9996        12-Jan-16  5587-7265-8118-7718              49957   
9997         8-Dec-16  5587-7265-8118-7718              13508   
9998         5-Feb-16  5587-7265-8118-7718                624   
9999         5-Apr-16  5587-7265-8118-7718              14019   

     transaction_segment  
0                  SEG21  
1                  SEG12  
2                  SEG11  
3                  SEG11  
4                  SEG18  
...                  ...  
9995               SEG15  
9996               SEG18  
9997               SEG25  
9998               SEG24  
9999               SEG13  

[10000 rows x 9 columns]

## Tidy Data

### Question 5

Read in hat_observations and split `observation` into two columns, `hat_colour` and `hat_type`.

Hint: Look up the pandas documentation for `.str.split()`

In [17]:
hat_observations = pd.read_csv("data/hat_observations.csv").clean_names()

In [18]:
hat_observations

day  month  year        observation  observation_count
0    25     10  2019        black,beret                  7
1    11      9  2019       blue,top hat                 10
2     4     12  2018    green,santa hat                  9
3    18      6  2018        white,beret                  9
4    26      9  2018    black,santa hat                  1
..  ...    ...   ...                ...                ...
95   15      8  2018  purple,bucket hat                  8
96   26      1  2018   green,bowler hat                  5
97   15     11  2018      black,top hat                  7
98   14      9  2019       green,helmet                  4
99    1     10  2018        blue,fedora                  4

[100 rows x 5 columns]

In [19]:
hat_observations[['hat_colour', 'hat_type']] = hat_observations['observation'].str.split(',', expand=True)

In [20]:
hat_observations

day  month  year        observation  observation_count hat_colour  \
0    25     10  2019        black,beret                  7      black   
1    11      9  2019       blue,top hat                 10       blue   
2     4     12  2018    green,santa hat                  9      green   
3    18      6  2018        white,beret                  9      white   
4    26      9  2018    black,santa hat                  1      black   
..  ...    ...   ...                ...                ...        ...   
95   15      8  2018  purple,bucket hat                  8     purple   
96   26      1  2018   green,bowler hat                  5      green   
97   15     11  2018      black,top hat                  7      black   
98   14      9  2019       green,helmet                  4      green   
99    1     10  2018        blue,fedora                  4       blue   

      hat_type  
0        beret  
1      top hat  
2    santa hat  
3        beret  
4    santa hat  
..         ...  
95  bucket hat  
96  bowler hat  
97     top hat  
98      helmet  
99      fedora  

[100 rows x 7 columns]

### Question 6

Unite day, month, and year columns into a column called date using a suitable separator. Then find the date where
the most berets were observed.

Hints:
- you can use the `.agg()` function to combine columns
- you may need to convert the data types of the columns you wish to combine

In [22]:
hat_observations.dtypes

day                   int64
month                 int64
year                  int64
observation          object
observation_count     int64
hat_colour           object
hat_type             object
dtype: object

In [25]:
hat_observations['date'] = pd.to_datetime(hat_observations[['year', 'month', 'day']])

In [26]:
hat_observations

day  month  year        observation  observation_count hat_colour  \
0    25     10  2019        black,beret                  7      black   
1    11      9  2019       blue,top hat                 10       blue   
2     4     12  2018    green,santa hat                  9      green   
3    18      6  2018        white,beret                  9      white   
4    26      9  2018    black,santa hat                  1      black   
..  ...    ...   ...                ...                ...        ...   
95   15      8  2018  purple,bucket hat                  8     purple   
96   26      1  2018   green,bowler hat                  5      green   
97   15     11  2018      black,top hat                  7      black   
98   14      9  2019       green,helmet                  4      green   
99    1     10  2018        blue,fedora                  4       blue   

      hat_type       date  
0        beret 2019-10-25  
1      top hat 2019-09-11  
2    santa hat 2018-12-04  
3        beret 2018-06-18  
4    santa hat 2018-09-26  
..         ...        ...  
95  bucket hat 2018-08-15  
96  bowler hat 2018-01-26  
97     top hat 2018-11-15  
98      helmet 2019-09-14  
99      fedora 2018-10-01  

[100 rows x 8 columns]

# Extension

## Joins

### Question 1
Can you join all 4 credit card datasets together so that you’re left with a dataset that looks like below with 109 rows and 12 columns?

![](images/all_joined.png)

In [27]:
card_cust = card_base.merge(customer_base, how='inner', on='cust_id')

In [28]:
fraud_trans = fraud_base.merge(transaction_base, how='inner', on='transaction_id')

In [29]:
card_cust

card_number card_family  credit_limit  cust_id  age  \
0    8638-5407-3631-8196     Premium        530000  CC67088   27   
1    7106-4239-7093-1515        Gold         18000  CC12076   48   
2    6492-5655-8241-3530     Premium        596000  CC97173   22   
3    2868-5606-5152-5706        Gold         27000  CC55858   30   
4    1438-6906-2509-8219    Platinum        142000  CC90518   50   
..                   ...         ...           ...      ...  ...   
495  4708-4407-9601-6022     Premium        691000  CC64993   34   
496  6042-2856-7280-2925        Gold         33000  CC26787   46   
497  8706-3809-5167-3899     Premium        144000  CC32532   28   
498  9540-8558-5897-5046     Premium        830000  CC90246   35   
499  5587-7265-8118-7718        Gold         24000  CC37803   21   

    customer_segment customer_vintage_group  
0            Diamond                    VG1  
1               Gold                    VG3  
2            Diamond                    VG1  
3            Diamond                    VG1  
4           Platinum                    VG2  
..               ...                    ...  
495          Diamond                    VG1  
496         Platinum                    VG2  
497         Platinum                    VG2  
498             Gold                    VG3  
499          Diamond                    VG1  

[500 rows x 7 columns]

In [30]:
fraud_trans

transaction_id  fraud_flag transaction_date       credit_card_id  \
0     CTID50558449           1         6-May-16  4246-1369-3659-8804   
1     CTID55936882           1        29-Nov-16  1336-9200-1264-2551   
2     CTID63762180           1         5-Dec-16  8528-6154-7390-5081   
3     CTID76723439           1        15-Sep-16  7908-2695-7391-7499   
4     CTID21246201           1        29-Feb-16  2524-4184-5908-6750   
..             ...         ...              ...                  ...   
104   CTID57993591           1         2-Apr-16  6697-9358-9213-4896   
105   CTID91108283           1        17-Jul-16  9207-1270-6690-4905   
106   CTID30494187           1        31-Dec-16  9018-5320-5729-5393   
107   CTID51301522           1        28-Jan-16  3620-5235-2101-3391   
108   CTID49517337           1        30-Sep-16  9604-6821-2861-8084   

     transaction_value transaction_segment  
0                 6984               SEG23  
1                34367               SEG21  
2                44550               SEG15  
3                48275               SEG16  
4                35751               SEG25  
..                 ...                 ...  
104              31486               SEG24  
105              36706               SEG11  
106              14586               SEG20  
107              22354               SEG23  
108              35688               SEG24  

[109 rows x 6 columns]

In [31]:
card_cust_fraud_trans = card_cust.merge(fraud_trans, how='inner', left_on='card_number', right_on='credit_card_id')

In [32]:
card_cust_fraud_trans

card_number card_family  credit_limit  cust_id  age  \
0    7214-4915-6387-5429    Platinum        172000  CC72302   22   
1    6239-8641-8524-9441        Gold         26000  CC18438   41   
2    6228-9033-3508-8121        Gold         33000  CC36771   50   
3    6228-9033-3508-8121        Gold         33000  CC36771   50   
4    1466-1600-9312-7604     Premium        249000  CC60926   24   
..                   ...         ...           ...      ...  ...   
104  1947-8327-3848-6581    Platinum        113000  CC11690   24   
105  9018-5320-5729-5393     Premium        444000  CC54582   29   
106  2949-1972-1300-6680        Gold         45000  CC68567   31   
107  5384-8278-5616-5502     Premium        361000  CC83124   35   
108  9542-6082-9630-9970    Platinum        159000  CC66351   22   

    customer_segment customer_vintage_group transaction_id  fraud_flag  \
0            Diamond                    VG1   CTID87470159           1   
1            Diamond                    VG1   CTID40881434           1   
2            Diamond                    VG1   CTID42980186           1   
3            Diamond                    VG1   CTID33699337           1   
4            Diamond                    VG1   CTID25962688           1   
..               ...                    ...            ...         ...   
104          Diamond                    VG1   CTID16281374           1   
105             Gold                    VG3   CTID30494187           1   
106         Platinum                    VG2   CTID68584964           1   
107             Gold                    VG3   CTID38931999           1   
108         Platinum                    VG2   CTID88342446           1   

    transaction_date       credit_card_id  transaction_value  \
0          11-Mar-16  7214-4915-6387-5429              32198   
1          10-Sep-16  6239-8641-8524-9441              43482   
2          17-Feb-16  6228-9033-3508-8121              45625   
3          16-Apr-16  6228-9033-3508-8121              14442   
4          18-Aug-16  1466-1600-9312-7604              45807   
..               ...                  ...                ...   
104         6-Sep-16  1947-8327-3848-6581              34290   
105        31-Dec-16  9018-5320-5729-5393              14586   
106        31-Dec-16  2949-1972-1300-6680              18851   
107        28-Sep-16  5384-8278-5616-5502              29079   
108        21-Apr-16  9542-6082-9630-9970              32930   

    transaction_segment  
0                 SEG21  
1                 SEG15  
2                 SEG22  
3                 SEG16  
4                 SEG20  
..                  ...  
104               SEG21  
105               SEG20  
106               SEG22  
107               SEG23  
108               SEG14  

[109 rows x 13 columns]

In [33]:
card_cust_fraud_trans = card_cust_fraud_trans.drop('credit_card_id', axis=1)

In [34]:
card_cust_fraud_trans

card_number card_family  credit_limit  cust_id  age  \
0    7214-4915-6387-5429    Platinum        172000  CC72302   22   
1    6239-8641-8524-9441        Gold         26000  CC18438   41   
2    6228-9033-3508-8121        Gold         33000  CC36771   50   
3    6228-9033-3508-8121        Gold         33000  CC36771   50   
4    1466-1600-9312-7604     Premium        249000  CC60926   24   
..                   ...         ...           ...      ...  ...   
104  1947-8327-3848-6581    Platinum        113000  CC11690   24   
105  9018-5320-5729-5393     Premium        444000  CC54582   29   
106  2949-1972-1300-6680        Gold         45000  CC68567   31   
107  5384-8278-5616-5502     Premium        361000  CC83124   35   
108  9542-6082-9630-9970    Platinum        159000  CC66351   22   

    customer_segment customer_vintage_group transaction_id  fraud_flag  \
0            Diamond                    VG1   CTID87470159           1   
1            Diamond                    VG1   CTID40881434           1   
2            Diamond                    VG1   CTID42980186           1   
3            Diamond                    VG1   CTID33699337           1   
4            Diamond                    VG1   CTID25962688           1   
..               ...                    ...            ...         ...   
104          Diamond                    VG1   CTID16281374           1   
105             Gold                    VG3   CTID30494187           1   
106         Platinum                    VG2   CTID68584964           1   
107             Gold                    VG3   CTID38931999           1   
108         Platinum                    VG2   CTID88342446           1   

    transaction_date  transaction_value transaction_segment  
0          11-Mar-16              32198               SEG21  
1          10-Sep-16              43482               SEG15  
2          17-Feb-16              45625               SEG22  
3          16-Apr-16              14442               SEG16  
4          18-Aug-16              45807               SEG20  
..               ...                ...                 ...  
104         6-Sep-16              34290               SEG21  
105        31-Dec-16              14586               SEG20  
106        31-Dec-16              18851               SEG22  
107        28-Sep-16              29079               SEG23  
108        21-Apr-16              32930               SEG14  

[109 rows x 12 columns]

## Tidy Data

### Question 2

Read in exam_scores and transform it into long format with two new columns `exam_question` and `score`. Remove 
superfluous information from the values in `exam_question`.

In [52]:
exam_scores = pd.read_csv("data/exam_scores.csv").clean_names()

In [36]:
exam_scores

id  exam_q1  exam_q2  exam_q3  exam_q4  exam_q5  exam_q6  exam_q7  \
0      1       69       59       95       56       54       78       86   
1      2       37       93       56       75       20       55       20   
2      3       38       23       47       37       92       26       73   
3      4       49       79       80       58       23       45       21   
4      5       40       92       59       47       30       66       53   
..   ...      ...      ...      ...      ...      ...      ...      ...   
195  196       75       71       98       87       30       46       91   
196  197       66       60       79       98       37       48       85   
197  198       99       66       67       22       68       35       65   
198  199       22       99       53       66       39       85       59   
199  200       65       86       30       21       85       74       83   

     exam_q8  exam_q9  exam_q10  
0         99       34        47  
1         93       75        70  
2         34       87        22  
3         51       25        56  
4         64       86        64  
..       ...      ...       ...  
195       40       43        54  
196       59       30        28  
197       21       60        28  
198       97       67        67  
199      100       90        33  

[200 rows x 11 columns]

In [53]:
exam_scores = exam_scores.melt(id_vars = "id", var_name = "exam_question", value_name = "score")

In [54]:
exam_scores

id exam_question  score
0       1       exam_q1     69
1       2       exam_q1     37
2       3       exam_q1     38
3       4       exam_q1     49
4       5       exam_q1     40
...   ...           ...    ...
1995  196      exam_q10     54
1996  197      exam_q10     28
1997  198      exam_q10     28
1998  199      exam_q10     67
1999  200      exam_q10     33

[2000 rows x 3 columns]